### DISCUSSION TITLE
Does this have feature parity with snoowrap?

### DISCUSSION THREAD

**Discussion by FoxxMD** [thislooksfun/snoots#42] [Q&A] [answered]
I've looked over the docs/migration section but nowhere does it mention if I can do everything in snoots that I can do in snoowrap. 

Even if it means refactoring to do things the "snoots" way, could I fullly convert a project using every feature in snoowrap to using snoots?

**Comment by thislooksfun** [comment, marked as answer]
# Short answer
Not yet, check #46 for progress

# Long answer

There are still several large areas of the codebase that need work, namely own-user methods, [private messaging][pmd], modmail, and flairs. For a full class-by-class rundown of what is missing see below.

## Snoowrap Classes:

#### Notes/caveats:
 - This is comparing `snoowrap@1.23.0` against `snoots@v1.0.0-dev.14`.
 - The names used here are `snoowrap`'s names. `snoots`' naming scheme is often different.
 - I tried to be as complete as possible but it's likely I missed something.
 - This does not cover any method *parameters*, just the methods themselves.

### [snoowrap](https://not-an-aardvark.github.io/snoowrap/snoowrap.html)
- Missing properties
  - `accessToken`<sup>1</sup>
  - `clientId`<sup>1</sup>
  - `clientSecret`<sup>1</sup>
  - `password`<sup>1</sup>
  - `ratelimitExpiration`
  - `ratelimitRemaining`
  - `scope`
  - `tokenExpiration`<sup>2</sup>
  - `userAgent`<sup>1</sup>
  - `username`<sup>1</sup>
- Missing methods
  - [`getMessage()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getMessage__anchor)<sup>3</sup>
  - [`getLivethread()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getLivethread__anchor)
  - [`getKarma()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getKarma__anchor)
  - [`getPreferences()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getPreferences__anchor)
  - [`updatePreferences()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#updatePreferences__anchor)
  - [`getMyTrophies()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getMyTrophies__anchor)
  - [`getFriends()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getFriends__anchor)
  - [`getBlockedUsers()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getBlockedUsers__anchor)
  - [`checkCaptchaRequirement()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#checkCaptchaRequirement__anchor)
  - [`getNewCaptchaIdentifier()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getNewCaptchaIdentifier__anchor)
  - [`getCaptchaImage()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getCaptchaImage__anchor)
  - [`getSavedCategories()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getSavedCategories__anchor)
  - [`markAsVisited()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#markAsVisited__anchor)
  - [`getBest()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getBest__anchor)
  - [`getContentByIds()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getContentByIds__anchor)<sup>2</sup>
  - [`getUnreadMessages()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getUnreadMessages__anchor)<sup>3</sup>
  - [`getInbox()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getInbox__anchor)<sup>3</sup>
  - [`getModmail()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getModmail__anchor)<sup>1</sup>
  - [`getNewModmailConversations()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getNewModmailConversations__anchor)
  - [`createModmailDiscussion()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#createModmailDiscussion__anchor)
  - [`getNewModmailConversation()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getNewModmailConversation__anchor)
  - [`markNewModmailConversationsAsRead()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#markNewModmailConversationsAsRead__anchor)
  - [`markNewModmailConversationsAsUnread()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#markNewModmailConversationsAsUnread__anchor)
  - [`getNewModmailSubreddits()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getNewModmailSubreddits__anchor)
  - [`getUnreadNewModmailConversationsCount()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getUnreadNewModmailConversationsCount__anchor)
  - [`bulkReadNewModmail()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#bulkReadNewModmail__anchor)
  - [`getSentMessages()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getSentMessages__anchor)<sup>3</sup>
  - [`markMessagesAsRead()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#markMessagesAsRead__anchor)<sup>3</sup>
  - [`markMessagesAsUnread()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#markMessagesAsUnread__anchor)<sup>3</sup>
  - [`readAllMessages()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#readAllMessages__anchor)<sup>3</sup>
  - [`composeMessage()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#composeMessage__anchor)<sup>3</sup>
  - [`getOauthScopeList()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getOauthScopeList__anchor)
  - [`searchSubredditNames()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#searchSubredditNames__anchor)
  - [`createSubreddit()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#createSubreddit__anchor)
  - [`searchSubredditTopics()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#searchSubredditTopics__anchor)
  - [`getSubscriptions()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getSubscriptions__anchor)
  - [`getContributorSubreddits()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getContributorSubreddits__anchor)
  - [`getModeratedSubreddits()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getModeratedSubreddits__anchor)
  - [`searchSubreddits()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#searchSubreddits__anchor)
  - [`getPopularSubreddits()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getPopularSubreddits__anchor)
  - [`getNewSubreddits()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getNewSubreddits__anchor)
  - [`getGoldSubreddits()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getGoldSubreddits__anchor)
  - [`getDefaultSubreddits()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getDefaultSubreddits__anchor)
  - [`checkUsernameAvailability()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#checkUsernameAvailability__anchor)
  - [`createLivethread()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#createLivethread__anchor)
  - [`getStickiedLivethread()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getStickiedLivethread__anchor)
  - [`getMyMultireddits()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#getMyMultireddits__anchor)
  - [`createMultireddit()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#createMultireddit__anchor)
  - [`revokeAccessToken()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#revokeAccessToken__anchor)
  - [`revokeRefreshToken()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#revokeRefreshToken__anchor)
  - [`noConflict()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#noConflict__anchor)<sup>1</sup>
  - [`oauthRequest()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#oauthRequest__anchor)<sup>1</sup>
  - [`credentialedClientRequest()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#credentialedClientRequest__anchor)<sup>1</sup>
  - [`unauthenticatedRequest()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#unauthenticatedRequest__anchor)<sup>1</sup>
  - [`rawRequest()`](https://not-an-aardvark.github.io/snoowrap/snoowrap.html#rawRequest__anchor)<sup>1</sup>

### [RedditContent](https://not-an-aardvark.github.io/snoowrap/RedditContent.html) (base class for all items)
- Missing properties
  - `created`<sup>1</sup>
- Misisng methods
  - [`fetch()`](https://not-an-aardvark.github.io/snoowrap/RedditContent.html#fetch__anchor)<sup>1</sup>
  - [`refresh()`](https://not-an-aardvark.github.io/snoowrap/RedditContent.html#refresh__anchor)<sup>1</sup>
  - [`toJSON()`](https://not-an-aardvark.github.io/snoowrap/RedditContent.html#toJSON__anchor)<sup>1</sup>

### [ReplyableContent](https://not-an-aardvark.github.io/snoowrap/ReplyableContent.html) (extends RedditContent)
Full feature parity.

### [VoteableContent](https://not-an-aardvark.github.io/snoowrap/VoteableContent.html) (extends ReplyableContent)
- Missing properties
  - `author_flair_background_color`
  - `author_flair_css_class`
  - `author_flair_richtext`
  - `author_flair_template_id`
  - `author_flair_text`
  - `author_flair_text_color`
  - `author_flair_type`
  - `author_fullname`<sup>1</sup>
  - `downs`<sup>1</sup>
  - `no_follow`
  - `removal_reason`
  - `report_reasons`<sup>2</sup>
  - `subreddit_id`<sup>1</sup>
  - `subreddit_name_prefixed`<sup>1</sup>
  - `ups`<sup>1</sup>

### [Comment](https://not-an-aardvark.github.io/snoowrap/Comment.html) (extends VoteableContent)
- Missing properties
  - `collapsed_reason`<sup>2</sup>
  - `controversiality`
  - `depth`<sup>2</sup>
- Missing methods
  - [`lock()`](https://not-an-aardvark.github.io/snoowrap/Comment.html#lock__anchor)
  - [`unlock()`](https://not-an-aardvark.github.io/snoowrap/Comment.html#unlock__anchor)
  - [`expandReplies()`](https://not-an-aardvark.github.io/snoowrap/Comment.html#expandReplies__anchor)<sup>1</sup>

### [RedditUser](https://not-an-aardvark.github.io/snoowrap/RedditUser.html) (extends RedditContent)
- Properties
  - `features`
  - `gold_creddits`
  - `modhash`<sup>2</sup>
  - `new_modmail_exists`<sup>1</sup>
  - `oauth_client_id`<sup>2</sup>
  - `pref_clickgadget`<sup>2</sup>
  - `pref_geopopular`<sup>2</sup>
  - `pref_show_snoovatar`<sup>2</sup>
  - `pref_top_karma_subreddits`<sup>2</sup>
  - `subreddit`
  - `subreddit`
- Methods
  - [`giveGold()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#giveGold__anchor)
  - [`assignFlair()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#assignFlair__anchor)
  - [`friend()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#friend__anchor)
  - [`unfriend()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#unfriend__anchor)
  - [`getFriendInformation()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#getFriendInformation__anchor)
  - [`getTrophies()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#getTrophies__anchor)
  - [`getOverview()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#getOverview__anchor)
  - [`getUpvotedContent()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#getUpvotedContent__anchor)
  - [`getDownvotedContent()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#getDownvotedContent__anchor)
  - [`getHiddenContent()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#getHiddenContent__anchor)
  - [`getSavedContent()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#getSavedContent__anchor)
  - [`getGildedContent()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#getGildedContent__anchor)
  - [`getMultireddit()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#getMultireddit__anchor)
  - [`getMultireddits()`](https://not-an-aardvark.github.io/snoowrap/RedditUser.html#getMultireddits__anchor)

### [Submission](https://not-an-aardvark.github.io/snoowrap/Submission.html) (extends VoteableContent)
- Missing properties
  - `clicked`
  - `content_categories`
  - `link_flair_background_color`
  - `link_flair_css_class`
  - `link_flair_richtext`
  - `link_flair_template_id`
  - `link_flair_text`
  - `link_flair_text_color`
  - `link_flair_type`
  - `media`
  - `media_embed`
  - `media_only`
  - `parent_whitelist_status`
  - `previous_visits`
  - `pwls`<sup>2</sup>
  - `post_hint`
  - `preview`
  - `quarantine`
  - `removal_reason`
  - `removed_by_category`
  - `secure_media`
  - `secure_media_embed`
  - `selftext`
  - `selftext_html`
  - `suggested_sort`
  - `thumbnail`
  - `view_count`
  - `visited`
  - `whitelist_status`
  - `wls`<sup>2</sup>
- Missing Methods
  - [`assignFlair()`](https://not-an-aardvark.github.io/snoowrap/Submission.html#assignFlair__anchor)
  - [`disableContestMode()`](https://not-an-aardvark.github.io/snoowrap/Submission.html#disableContestMode__anchor)
  - [`enableContestMode()`](https://not-an-aardvark.github.io/snoowrap/Submission.html#enableContestMode__anchor)
  - [`getLinkFlairTemplates()`](https://not-an-aardvark.github.io/snoowrap/Submission.html#getLinkFlairTemplates__anchor)
  - [`getRelated()`](https://not-an-aardvark.github.io/snoowrap/Submission.html#getRelated__anchor)<sup>1</sup>
  - [`markAsRead()`](https://not-an-aardvark.github.io/snoowrap/Submission.html#markAsRead__anchor)
  - [`selectFlair()`](https://not-an-aardvark.github.io/snoowrap/Submission.html#selectFlair__anchor)
  - [`setSuggestedSort()`](https://not-an-aardvark.github.io/snoowrap/Submission.html#setSuggestedSort__anchor)

### [LiveThread](https://not-an-aardvark.github.io/snoowrap/LiveThread.html) (extends RedditContent)
Unimplemented

### [PrivateMessage](https://not-an-aardvark.github.io/snoowrap/PrivateMessage.html) (extends ReplyableContent)
Unimplemented

### [Subreddit](https://not-an-aardvark.github.io/snoowrap/Subreddit.html) (extends RedditContent)
- Missing properties
  - `accounts_active`<sup>1</sup>
  - `advertiser_category`
  - `allow_discovery`
  - `allow_images`
  - `allow_videogifs`
  - `allow_videos`
  - `has_menu_widget`
  - `icon_img`
  - `icon_size`
  - `is_enrolled_in_new_modmail`<sup>1</sup>
  - `key_color`
  - `lang`
  - `link_flair_enabled`
  - `link_flair_position`
  - `notification_level`
  - `primary_color`
  - `public_description_html`
  - `public_description`
  - `suggested_comment_sort`
  - `user_can_flair_in_sr`
  - `user_flair_background_color`
  - `user_flair_css_class`
  - `user_flair_enabled_in_sr`
  - `user_flair_position`
  - `user_flair_richtext`
  - `user_flair_template_id`
  - `user_flair_text`
  - `user_flair_text_color`
  - `whitelist_status`
  - `wls`
- Missing methods
  - [`configureFlair()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#configureFlair__anchor)
  - [`createLinkFlairTemplate()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#createLinkFlairTemplate__anchor)
  - [`createUserFlairTemplate()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#createUserFlairTemplate__anchor)
  - [`deleteAllLinkFlairTemplates()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#deleteAllLinkFlairTemplates__anchor)
  - [`deleteAllUserFlairTemplates()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#deleteAllUserFlairTemplates__anchor)
  - [`deleteBanner()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#deleteBanner__anchor)
  - [`deleteFlairTemplate()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#deleteFlairTemplate__anchor)
  - [`deleteHeader()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#deleteHeader__anchor)
  - [`deleteIcon()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#deleteIcon__anchor)
  - [`deleteImage()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#deleteImage__anchor)
  - [`deleteUserFlair()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#deleteUserFlair__anchor)
  - [`editSettings()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#editSettings__anchor)
  - [`getBannedUsers()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getBannedUsers__anchor)
  - [`getContributors()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getContributors__anchor)
  - [`getLinkFlairTemplates()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getLinkFlairTemplates__anchor)
  - [`getModerationLog()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getModerationLog__anchor)
  - [`getModerators()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getModerators__anchor)
  - [`getModmail()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getModmail__anchor)
  - [`getNewModmailConversations()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getNewModmailConversations__anchor)
  - [`getMutedUsers()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getMutedUsers__anchor)
  - [`getMyFlair()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getMyFlair__anchor)
  - [`getRecommendedSubreddits()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getRecommendedSubreddits__anchor)
  - [`getRules()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getRules__anchor)
  - [`getSettings()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getSettings__anchor)
  - [`getSticky()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getSticky__anchor)
  - [`getStylesheet()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getStylesheet__anchor)
  - [`getSubmitText()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getSubmitText__anchor)
  - [`getUserFlair()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getUserFlair__anchor)
  - [`getUserFlairList()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getUserFlairList__anchor)
  - [`getUserFlairTemplates()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getUserFlairTemplates__anchor)
  - [`getWikiBannedUsers()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getWikiBannedUsers__anchor)
  - [`getWikiContributors()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getWikiContributors__anchor)
  - [`getWikiPage()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getWikiPage__anchor)
  - [`getWikiPages()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getWikiPages__anchor)
  - [`getWikiRevisions()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#getWikiRevisions__anchor)
  - [`hideMyFlair()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#hideMyFlair__anchor)
  - [`inviteModerator()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#inviteModerator__anchor)
  - [`leaveContributor()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#leaveContributor__anchor)<sup>1</sup>
  - [`leaveModerator()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#leaveModerator__anchor)<sup>1</sup>
  - [`muteUser()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#muteUser__anchor)
  - [`removeModerator()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#removeModerator__anchor)
  - [`revokeModeratorInvite()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#revokeModeratorInvite__anchor)
  - [`selectMyFlair()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#selectMyFlair__anchor)
  - [`setModeratorPermissions()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#setModeratorPermissions__anchor)
  - [`setMultipleUserFlairs()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#setMultipleUserFlairs__anchor)
  - [`showMyFlair()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#showMyFlair__anchor)
  - [`submitLink()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#submitLink__anchor)
  - [`submitSelfpost()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#submitSelfpost__anchor)
  - [`subscribe()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#subscribe__anchor)
  - [`unmuteUser()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#unmuteUser__anchor)
  - [`unsubscribe()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#unsubscribe__anchor)
  - [`unwikibanUser()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#unwikibanUser__anchor)
  - [`updateStylesheet()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#updateStylesheet__anchor)
  - [`uploadBannerImage()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#uploadBannerImage__anchor)
  - [`uploadHeaderImage()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#uploadHeaderImage__anchor)
  - [`uploadIcon()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#uploadIcon__anchor)
  - [`uploadStylesheetImage()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#uploadStylesheetImage__anchor)
  - [`wikibanUser()`](https://not-an-aardvark.github.io/snoowrap/Subreddit.html#wikibanUser__anchor)

### [MultiReddit](https://not-an-aardvark.github.io/snoowrap/MultiReddit.html) (extends RedditContent)
Unimplemented

### [ModmailConversation](https://not-an-aardvark.github.io/snoowrap/ModmailConversation.html) (extends RedditContent)
Unimplemented

### [ModmailConversationAuthor](https://not-an-aardvark.github.io/snoowrap/ModmailConversationAuthor.html) (extends RedditContent)
Unimplemented

### [WikiPage](https://not-an-aardvark.github.io/snoowrap/WikiPage.html) (extends RedditContent)
Unimplemented

### [Listing](https://not-an-aardvark.github.io/snoowrap/Listing.html)
Feature parity, but implemented very differently.

---

<sup>1</sup>This will not be implemented.
<sup>2</sup>This is unlikely to be implemented.
<sup>3</sup>This relates to private messaging, which is an [ongoing discussion][pmd].

[pmd]: https://github.com/thislooksfun/snoots/discussions/30

**Reply by FoxxMD** [reply]
Thank for you the detailed rundown. With regards to things that will not be implemented (1) I understand many of them are deprecated properties or endpoints from reddit or things snoowrap does that are weird in general. However what is your reasoning for not exposing oauth info and rate limit information?

In my application that uses snoowrap I rely heavily on being able to retrieve the scopes available from an authenticated client and rate limit information for managing api usage in real time. Are you planning on making these available in some other way such as hooks/middleware into request-response objects?

**Reply by thislooksfun** [reply]
Honestly I didn't think they would be useful for introspection.

The following are pure inputs (i.e. `snoowrap` just exposes the values you pass in), so you can keep track of these yourself:
- `clientId`
- `clientSecret`
- `password`
- `userAgent`
- `username`

These I did not expect to be useful to expose (and aren't currently being tracked, though that's wip), but I can definitely add some getters for them:
- `ratelimitExpiration`
- `ratelimitRemaining`
- `scope`

The token lifecycle is managed automatically, so I see no logical use for exposing `accessToken` and `tokenExpiration`.

And finally `refreshToken` actually *is* already implemeted via [`Client#getRefreshToken()`](https://thislooks.fun/snoots/docs/latest/classes/client.html#getrefreshtoken) (my bad!)

I have updated my previous comment accordingly.

If you have any more questions please don't hesitate to reach out!

**Reply by FoxxMD** [reply]
Thank you for re-considering on those points. To give you a better idea of how I use them (as a single use-case) and why they'd be important:

### Rate Limit

* How my application responds to a 4xx response from reddit is different based on whether it is due to rate limit exhaustion or some other reason. For any other reason I have a have retry with exponential falloff but for rate limit current behavior is immediately stopped.
* I sample rate limit usage and use it in conjunction with expiration to...
   * Show users real-time quota IE `577/600 remaining` and a real-time estimate of depletion time IE `in ~5 minutes`
   * [Have soft/hard thresholds based on both actual # remaining and expected depletion](https://github.com/FoxxMD/context-mod/blob/master/src/Bot/index.ts#L682)  -- when soft threshold is hit i reduce polling rates and response rates to try to stay within limit without stopping all activity
  * [Based on rate limit usage per subreddit](https://github.com/FoxxMD/context-mod/blob/master/src/Bot/index.ts#L765) I turn off some subreddit behavior before others (high usage is turned off first)
* (WIP) Showing graphed usage of api usage per behavior/subreddit to clients

### Scope

After making a test call snoowrap updates scope for the current authenticated user when it refreshes the token. I then makes checks against scope and mod permissions to determine if my bot can perform some actions like...
 
* The bot must have bot mod permission `Manager Flairs` **and the oauth scope** `flair` in order to set flair on users/posts.
* The bot must have `modself` oauth permission in order to check for and accept moderator invites. I check this before enabling programmatic invitation management per bot as well as giving the user a hint the bot must have it.

# Requests

One other feature I'd like to touch on is access to http requests. 

### Exposing Raw Authenticated Request/Response

It looks like the main snoots client does expose `get` and `post` "raw" authenticated calls so that does kind of cover `oauthRequest()` which you have marked as not implementing (🤔 ). If those are actually exposed and raw responses are available that's all good 👍  

Having access to those means that if there is something snoots has not implemented, or has not implemented the way I would need, I can always [extend the client with my own method](https://github.com/FoxxMD/context-mod/blob/master/src/Utils/SnoowrapClients.ts#L17) and use those functions to cover that. Just don't make them private ;)

### Client Extensibility

You also have `rawRequest()` marked as not implementing. Understandable but I'd urge you to consider exposing `got` to the user so we can take full advantage of [hooks, events, and plugins](https://github.com/sindresorhus/got/blob/main/readme.md#advanced-creation). 

One use-case for this, that I already use a la [overriding `rawRequest` in snoowrap](https://github.com/FoxxMD/context-mod/blob/master/src/Utils/SnoowrapClients.ts#L63), is configuring the requests to route through a local proxy server so that I can simulate network issues and 4xx/5xx responses from reddit to my application. If you just expose `got` to the user we can implement this kind of behavior ourselves without anymore work on your end.

**Reply by FoxxMD** [reply]
Realizing this should probably be in #4 instead of here? Should I move my comments to there?

**Reply by thislooksfun** [reply]
From #4:

> To make tracking these easier, please make separate threads for each piece of feedback.

So no, it's actually better here.

As for your reasoning I haven't had time to respond in detail, but I will get to it either later tonight or tomorrow.

**Reply by thislooksfun** [reply]
@FoxxMD Alright, I've now had some time to review what you said. Thank you for the insight into how these fields are currently being used with snoowrap. You make some excellent points, and I will say that I definitely want to make usage like that possible with snoots.

> How my application responds to a 4xx response from reddit is different based on whether it is due to rate limit exhaustion or some other reason.

That makes perfect sense. For a while now I have been planning to add a feature to snoots that mirrors snoowrap `continueAfterRatelimitError` flag. I could also add a built-in exponential retry for certain classes of error if that would be useful.
  
> * Show users real-time quota IE `577/600 remaining` and a real-time estimate of depletion time IE `in ~5 minutes`
> * Have soft/hard thresholds based on both actual # remaining and expected depletion
> * Turn off some subreddit behavior before others
> * Showing graphed usage of api usage per behavior/subreddit to clients

All fantastic ideas, all will be possible.

> snoowrap updates scope for the current authenticated user when it refreshes the token

Yes, I noticed that. Currently snoots is just [discarding the scope(s)](https://github.com/thislooksfun/snoots/blob/a2e0f6b98b4ec98e5d5e66db2428f318a4014785/src/helper/accessToken.ts#L13-L28), but a future version will definitely expose that (see #43 for progress).

> The main snoots client does expose `get` and `post` "raw" authenticated calls so that does kind of cover `oauthRequest()` which you have marked as not implementing (🤔 ).

Those raw methods are only going to be exposed until snoots reaches feature parity with the reddit api. That being said, once that happens they will only be made `protected`, never `private`, so you can still freely extend the class to use them directly should you absolutely need to (though it would be greatly appreciated if you'd open an issue / PR so we can all benefit!).

As for why I said I wouldn't implement it, snoowrap exposes separate `oauthRequest()`, `credentialedClientRequest()` and `unauthenticatedRequest()` methods. Snoots has the same three code paths, but the switching between them is handled internally based on the authentication state (`oauth` if it has a token, `credentialedClient` if it only has creds, and `unauthenticated` if it has neither). If you need to manually switch between these auth levels using snoots you need to make at least 3 clients, one for each desired auth level.

Additionally, I am now planning a refactor of the internal request framework to make it more structured, more extensible, and more mockable. This refactor will allow for all the super-low-level inspection you want (including entirely bypassing any built-in error handling and retry mechanisms, should that be necessary). More details soon, still gotta iron out exactly how I want to do it.

Thanks again for the feedback!

**Reply by thislooksfun** [reply]
@FoxxMD Just FYI, I will be tracking progress towards feature parity in #46, so go bookmark that for future updates.
